libs/template/renderer.go

```diff
@@ -0,0 +1,260 @@
+	// generate the project
+	templateFiler filer.Filer
+
+	// Filer rooted at the project instance root. Allows writing files generated
+	// when walking the template file tree
+	instanceFiler filer.Filer
+}
+
+func newRenderer(ctx context.Context, config map[string]any, libraryRoot, instanceRoot, templateRoot string) (*renderer, error) {
+	// Initialize new template, with helper functions loaded
+	tmpl := template.New("").Funcs(HelperFuncs)
+
+	// Load user defined associated templates from the library root
+	libraryGlob := filepath.Join(libraryRoot, "*")
+	matches, err := filepath.Glob(libraryGlob)
+	if err != nil {
+		return nil, err
+	}
+	if len(matches) != 0 {
+		tmpl, err = tmpl.ParseGlob(libraryGlob)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	templateFiler, err := filer.NewLocalClient(templateRoot)
+	if err != nil {
+		return nil, err
+	}
+
+	instanceFiler, err := filer.NewLocalClient(instanceRoot)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx = log.NewContext(ctx, log.GetLogger(ctx).With("action", "initialize-template"))
+
+	return &renderer{
+		ctx:           ctx,
+		config:        config,
+		baseTemplate:  tmpl,
+		files:         make([]*inMemoryFile, 0),
+		skipPatterns:  make([]string, 0),
+		templateFiler: templateFiler,
+		instanceFiler: instanceFiler,
+	}, nil
+}
+
+// Executes the template by applying config on it. Returns the materialized template
+// as a string
+func (r *renderer) executeTemplate(templateDefinition string) (string, error) {
+	// Create copy of base template so as to not overwrite it
+	tmpl, err := r.baseTemplate.Clone()
+	if err != nil {
+		return "", err
+	}
+
+	// Parse the template text
+	tmpl, err = tmpl.Parse(templateDefinition)
+	if err != nil {
+		return "", err
+	}
+
+	// Execute template and get result
+	result := strings.Builder{}
+	err = tmpl.Execute(&result, r.config)
+	if err != nil {
+		return "", err
+	}
+	return result.String(), nil
+}
+
+func (r *renderer) computeFile(relPathTemplate string) (*inMemoryFile, error) {
+	// read template file contents
+	templateReader, err := r.templateFiler.Read(r.ctx, relPathTemplate)
+	if err != nil {
+		return nil, err
+	}
+	contentTemplate, err := io.ReadAll(templateReader)
+	if err != nil {
+		return nil, err
+	}
+
+	// execute the contents of the file as a template
+	content, err := r.executeTemplate(string(contentTemplate))
+	// Capture errors caused by the "fail" helper function
+	if target := (&ErrFail{}); errors.As(err, target) {
+		return nil, target
+	}
+	if err != nil {
+		return nil, fmt.Errorf("failed to compute file content for %s. %w", relPathTemplate, err)
+	}
+
+	// Execute relative path template to get materialized path for the file
+	relPath, err := r.executeTemplate(relPathTemplate)
+	if err != nil {
+		return nil, err
+	}
+
+	return &inMemoryFile{
+		path:    relPath,
+		content: []byte(content),
+	}, nil
+}
+
+func (r *renderer) walk() error {
+	directories := []string{"."}
```

Comment:
Can't you use WalkDir instead?

Reply:
Please add this as a comment to this function. I suspect others reading this code may think the same.

Reply:
done